File: ruby/numbers/src/numbers_recursive.rb
module NumbersRecursive
  def self.fibonacci(n)
    if n <= 0
      0
    elsif n == 1
      1
    else
      fibonacci(n - 1) + fibonacci(n - 2)
    end
  end

  def self.factorial(n)
    if n <= 1
      1
    else
      n * factorial(n - 1)
    end
  end

  def self.sum_numbers(n)
    if n <= 0
      0
    else
      n + sum_numbers(n - 1)
    end
  end
end

File: fortran/numbers/src/numbers_recursive.f90
module numbers_recursive
contains

  ! Recursive Fibonacci
  recursive function fibonacci(N) result(res)
    integer, intent(in) :: N
    integer :: res
    if (N <= 1) then
      res = N
    else
      res = fibonacci(N - 1) + fibonacci(N - 2)
    end if
  end function fibonacci

  ! Recursive factorial
  recursive function factorial(N) result(res)
    integer, intent(in) :: N
    integer :: res
    if (N == 1) then
      res = 1
    else
      res = N * factorial(N - 1)
    end if
  end function factorial

  ! Recursive sum of first N numbers
  recursive function sum_numbers(N) result(res)
    integer, intent(in) :: N
    integer :: res
    if (N == 0) then
      res = 0
    else
      res = N + sum_numbers(N - 1)
    end if
  end function sum_numbers

end module numbers_recursive